server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseActions.java

```diff
@@ -84,10 +87,14 @@
 
         @Override
         protected ShardsIterator shards(final ClusterState state, final InternalRequest request) {
-            return state
+            final IndexShardRoutingTable shardRoutingTable = state
                     .routingTable()
-                    .shardRoutingTable(request.concreteIndex(), request.request().getShardId().id())
-                    .primaryShardIt();
+                    .shardRoutingTable(request.concreteIndex(), request.request().getShardId().id());
+            if (shardRoutingTable.primaryShard().active()) {
```

Comment:
TransportSingleShardAction doesn't do our usual "chase the shard" pattern where we re-resolve shards on each node until we find a place where the shard is locally available. This means that if the coordinating node thinks the shard is active but the node with the shard didn't yet process the shard activation cluster state, I think this still goes wrong (i.e., the primary would not be in primary mode, which is [activated](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java#L482) when the cluster state is processed). I hope I'm wrong and please let me know what I'm missing.

Reply:
Good catch. In this case, I think that we should use the other approach that I [considered](
https://github.com/elastic/elasticsearch/pull/40386#issuecomment-475794196). I can not think of any situations where we would want to acquire a permit on a non-active primary?

Reply:
That was actually why I went looking - my instinct was that this should be done under permit and that the permit shouldn't be given under a non-initialized primary. We currently don't do that so that requires a much bigger change/vision. We can also have a targeted-check in asyncShardOperation that the shard is active before performing the operation.  @ywelsch any thoughts?

Reply:
`RetentionLeaseActions` and `TransportForgetFollowerAction` can also possibly violate the assertion in `acquirePrimaryOperationPermit` that the shard is actually a primary (by the time  the request arrives, the primary could have failed and a replica allocated instead). Ensuring this was previously left to the caller of this method.

We can explore changing `acquirePrimaryOperationPermit` and `acquireAllPrimaryOperationsPermits` to throw appropriate exceptions if the shard is not in primary mode (e.g. replica or initializing/relocated primary). TRA can then react to an `IndexShardRelocatedException` to delegate to the relocation target.

Reply:
That takes it one level higher - ensuring that the primary is actually an active primary (rather than just an active shard). SGTM.